scripts/add_electricity.py

```diff
@@ -286,6 +273,67 @@ def update_transmission_costs(n, costs, length_factor=1.0, simple_hvdc_costs=Fal
             + costs.at["HVDC inverter pair", "capital_cost"]
         )
     n.links.loc[dc_b, "capital_cost"] = costs
+    return n
+
+
+def attach_dc_costs_lines(n, costs, length_factor=1.0, simple_hvdc_costs=False):
```

Comment:
I really like the approach of combining the two links and lines into the same update_transmission_costs, so that we don't lose retrocompatibility of functionalities with respect to pypsa-earth :)

As a comment, at first glance, it seems that the _lines and _links function differ only by the "n.links" or "n.lines", is it right?
if the code works with that, that's an amazing flow that you found!
Because we could create an utility function such as "attach_dc_costs", general for both links and lines, that instead of taking n as first argument, it may be lines_or_link to represent n.links or n.lines.
In the body of the attach_dc_costs, the "n.links" may be replaced with lines_or_links and it (hopefully) work.

ideally, we could launch the function attach_dc_costs twice in the update_transmission_costs: once for the links e.g. `attach_dc_costs(n.links, ...)` and one for lines `attach_dc_costs(n.lines, ...)`

Within attach_dc_costs, then a check like `lines_or_links.loc[lines_or_links.carrier == "DC"].empty` may be checked to check if there are data to change, if not a simple return may apply similarly to before. Moving this check here, may clean also the code in update_transmisssion_costs that rightfully you did the check externally.

If the above is possible, that would be a dream, but I am unsure that it is feasible.
I think this strong work that you did may answer to this doubt :)

Reply:
Amazing idea! Exactly what I have been looking for :D 
Thank you so much!

Reply:
As far as I understand, this fragment simply sums costs for different parts of HVDC grid meaning that there should not be any difference between "as-links" and "as-lines" representation. Correct me please if I'm missing something

Have implemented and it seems to work fine. The "passed-by-assignment" magic is very handy :)